fe/fe-core/src/main/java/com/starrocks/sql/common/MetaUtils.java

```diff
@@ -142,12 +142,54 @@ public static Table getTable(ConnectContext session, TableName tableName) {
         }
         Table table = session.getGlobalStateMgr().getMetadataMgr().getTable(tableName.getCatalog(),
                 tableName.getDb(), tableName.getTbl());
+
+        if (table == null) {
+            throw new SemanticException("Table %s is not found", tableName.toString());
+        }
+        return table;
+    }
+
+    // get table by tableName, unlike getTable, this interface is session aware,
+    // which means if there is a temporary table with the same name,
+    // use temporary table first, otherwise, treat it as a permanent table
+    public static Table getSessionAwareTable(ConnectContext session, TableName tableName) {
+        if (Strings.isNullOrEmpty(tableName.getCatalog())) {
+            tableName.setCatalog(session.getCurrentCatalog());
+        }
+        Database database = getDatabase(session, tableName);
+
+        Table table = session.getGlobalStateMgr().getMetadataMgr().getTemporaryTable(
+                session.getSessionId(), tableName.getCatalog(), database.getId(), tableName.getTbl());
+        if (table != null) {
+            return table;
+        }
+        table = session.getGlobalStateMgr().getMetadataMgr().getTable(
+                tableName.getCatalog(), tableName.getDb(), tableName.getTbl());
         if (table == null) {
             throw new SemanticException("Table %s is not found", tableName.toString());
         }
         return table;
     }
 
+    public static Table getTable(ConnectContext session, Database database, TableName tableName) {
+        if (Strings.isNullOrEmpty(tableName.getCatalog())) {
+            tableName.setCatalog(session.getCurrentCatalog());
+        }
+        Table table = null;
+        // for internal catalog, try to use temporary table first
+        if (CatalogMgr.isInternalCatalog(tableName.getCatalog())) {
+            table = session.getGlobalStateMgr().getMetadataMgr().getTemporaryTable(
```

Comment:
also get tmp table first? what's the difference with `getSessionAwareTable()`?

Reply:
> also get tmp table first? what's the difference with `getSessionAwareTable()`?

The input argument of `getSessionAwareTable` is TableName, it should getDatabase first internally. 
but in many cases, we already get Database before getTable, so it's better to have an interface with Database as argument, so that we can avoid some overhead.